test/apps/iodemo/io_demo.cc

```diff
@@ -681,52 +696,56 @@ class DemoClient : public P2pDemoCommon {
 
     DemoClient(const options_t& test_opts) :
         P2pDemoCommon(test_opts),
-        _num_sent(0), _num_completed(0), _status(OK), _start_time(get_time()),
-        _retry(0), _read_callback_pool(opts().iomsg_size)
-    {
-        _status_str[OK]                    = "ok";
-        _status_str[ERROR]                 = "error";
-        _status_str[RUNTIME_EXCEEDED]      = "run-time exceeded";
-        _status_str[CONN_RETRIES_EXCEEDED] = "connection retries exceeded";
+        _num_sent(0), _num_completed(0),
+        _status(OK), _start_time(get_time()),
+        _retry(0), _read_callback_pool(opts().iomsg_size) {
     }
 
     typedef enum {
         OK,
-        ERROR,
-        RUNTIME_EXCEEDED,
-        CONN_RETRIES_EXCEEDED
+        CONN_RETRIES_EXCEEDED,
+        WAIT_RESPONSES_TIMEOUT
     } status_t;
 
-    size_t do_io_read(UcxConnection *conn, uint32_t sn) {
-        size_t data_size = get_data_size();
-        IoMessage *m     = _io_msg_pool.get();
+    uint32_t get_server_index(const UcxConnection *conn) {
+        return _conn_idx[conn];
+    }
+
+    size_t do_io_read(int server_index, uint32_t sn) {
+        server_info_t& server_info = _server_info[server_index];
+        size_t data_size           = get_data_size();
+        IoMessage *m               = _io_msg_pool.get();
 
         m->init(IO_READ, sn, data_size, opts().validate);
-        if (!send_io_message(conn, m)) {
-            return data_size;
+        if (!send_io_message(server_info.conn, m)) {
```

Comment:
can `server_info.conn` be NULL here?

Reply:
no, there is assert in line 994:
`            assert(_server_info[server_index].conn != NULL);`